src/network/core/os_abstraction.cpp

```diff
@@ -0,0 +1,184 @@
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @file os_abstraction.cpp OS specific implementations of functions of
+ *                          the OS abstraction layer for network stuff.
+ *
+ * The general idea is to have simple abstracting functions for things that
+ * require different implementations for different environments.
+ * In here the functions, and their documentation, are defined only once
+ * and the implementation contains the #ifdefs to change the implementation.
+ * Since Windows is usually different that is usually the first case, after
+ * that the behaviour is usually Unix/BSD-like with occasional variation.
+ */
+
+#include "stdafx.h"
+#include "os_abstraction.h"
+#include "../../string_func.h"
+
+#include "../../safeguards.h"
+
+/**
+ * Construct the network error with the given error code.
+ * @param error The error code.
+ */
+NetworkError::NetworkError(int error) : error(error)
+{
+}
+
+/**
+ * Construct the network error with the last available error.
+ */
+NetworkError::NetworkError()
+{
+#if defined(_WIN32)
+	this->error = WSAGetLastError();
+#elif defined(__OS2__)
+	this->error = sock_errno();
+#else
+	this->error = errno;
+#endif
+}
+
+/**
+ * Check whether this error describes that the operation would block.
+ * @return True iff the operation would block.
+ */
+bool NetworkError::WouldBlock() const
+{
+#if defined(_WIN32)
+	return this->error == WSAEWOULDBLOCK;
+#elif EWOULDBLOCK != EAGAIN
+	/* Usually EWOULDBLOCK and EAGAIN are the same, but sometimes they are not
+	 * and the POSIX.1 specification states that either should be checked. */
+	return this->error == EWOULDBLOCK || this->error == EAGAIN;
+#else
+	return this->error == EWOULDBLOCK;
+#endif
+}
+
+/**
+ * Check whether this error describes a connection reset.
+ * @return True iff the connection is reset.
+ */
+bool NetworkError::IsConnectionReset() const
+{
+#if defined(_WIN32)
+	return this->error == WSAECONNRESET;
+#else
+	return this->error == ECONNRESET;
+#endif
+}
+
+/**
+ * Check whether this error describes a connect is in progress.
+ * @return True iff the connect is already in progress.
+ */
+bool NetworkError::IsConnectInProgress() const
+{
+#if defined(_WIN32)
+	return this->error == WSAEWOULDBLOCK;
+#else
+	return this->error == EINPROGRESS;
+#endif
+}
+
+/**
+ * Get the C-string representation of the error message.
+ * @return The C-string representation that will get overwritten by next calls.
+ */
+const char *NetworkError::AsString() const
+{
+#if defined(_WIN32)
+	static char buffer[512];
```

Comment:
Potential thread safety issues here? Or all good because Network stuff always done from single thread?

Reply:
Potential thread safety issue... definitely. Though, bear in mind that strerror is equally thread-unsafe. So practically it's not unsafer. Since the code is not new to this patch, it and the strerror case should probably be done in a separate patch.

Reply:
I have made an attempt to make it thread safer, but I'm not really happy with the non-Windows situation. However solving that properly might become a major pain in the proverbial arse as the comment I placed might explain.